mk3/launchpad.go

```diff
@@ -0,0 +1,171 @@
+		for {
+
+			if !l.Block {
+				// sleep for a while before the new polling tick,
+				// otherwise operation is too intensive and blocking
+				time.Sleep(10 * time.Millisecond)
+				hits, err := pad.Read()
+				if err != nil {
+					continue
+				}
+				for i := range hits {
+					ch <- hits[i]
+				}
+			} else {
+				pad.Read() // Throw away whilst we're blocked.
+			}
+		}
+	}(l, ch)
+	return ch
+}
+
+func (l *Launchpad) BlockKeys(block bool) {
+	l.Block = block
+}
+
+func (l *Launchpad) IsBlocked() bool {
+	return l.Block
+}
+
+// Read reads hits from the input stream. It returns max 64 hits for each read.
+func (l *Launchpad) Read() (hits []Hit, err error) {
+	var evts []portmidi.Event
+	if evts, err = l.inputStream.Read(1024); err != nil {
+		return
+	}
+	for _, evt := range evts {
+		if evt.Data2 > 0 {
+			var x, y int64
+
+			x = evt.Data1%10 - 1
+			y = (8 - (evt.Data1-x)/10)
+
+			hits = append(hits, Hit{X: int(x), Y: int(y)})
+		}
+	}
+	return
+}
+
+// Light lights the button at x,y with the given red, green, and blue values.
+// All available colors are documented and visualized at Launchpad's Programmers Guide.
+func (l *Launchpad) Light(x, y, red int, green int, blue int) error {
+	// TODO(jbd): Support top row.
+	led := int64((8-y)*10 + x + 1)
+	//return l.outputStream.WriteShort(0x90, led, int64(color))
+	return l.outputStream.WriteSysExBytes(portmidi.Time(), []byte{0xF0, 0x00, 0x20, 0x29, 0x02, 0x0D, 0x03, 0x03, byte(led), byte(red), byte(green), byte(blue), 0xF7})
+}
+
+func (l *Launchpad) FlashLight(x int, y int, onColor int, offColor int) error {
+	// TODO(jbd): Support top row.
+	led := int64((8-y)*10 + x + 1)
+	return l.outputStream.WriteSysExBytes(portmidi.Time(), []byte{0xF0, 0x00, 0x20, 0x29, 0x02, 0x0D, 0x03, 0x01, byte(led), byte(onColor), byte(offColor), 0xF7})
+}
+
+// Reset turns off all buttons.
+func (l *Launchpad) Reset() error {
+	// Sends a "light all ligts" SysEx command with 0 color.
+	return l.outputStream.WriteSysExBytes(portmidi.Time(), []byte{0xf0, 0x00, 0x20, 0x29, 0x02, 0x18, 0x0e, 0x00, 0xf7})
+}
+
+// Set Programmers Mode.
+func (l *Launchpad) Program() error {
+	// Set programmer mode using a SysEx command.
+	return l.outputStream.WriteSysExBytes(portmidi.Time(), []byte{0xf0, 0x00, 0x20, 0x29, 0x02, 0x0d, 0x0e, 0x01, 0xf7})
+}
+
+func (l *Launchpad) Close() error {
+	l.inputStream.Close()
+	l.outputStream.Close()
+	return nil
+}
+
+// discovers the currently connected Launchpad device
+// as a MIDI device.
+func discover() (input portmidi.DeviceID, output portmidi.DeviceID, err error) {
+	in := -1
+	out := -1
+	for i := 0; i < portmidi.CountDevices(); i++ {
+		info := portmidi.Info(portmidi.DeviceID(i))
+		fmt.Printf("%s\n", info.Name)
+		if strings.Contains(info.Name, "Launchpad Mini MK3") {
+			if info.IsInputAvailable {
+				in = i
+			}
+			if info.IsOutputAvailable {
+				out = i
+			}
+		}
+	}
+	if in == -1 || out == -1 {
+		err = errors.New("launchpad: no launchpad is connected")
+	} else {
```

Comment:
Remove the else.

Reply:
comes from your original code. i think you need the else to be there to set the correct values if there is indeed a launchpad connnected.